### PR TITLE
Feature/add tree information to profiler

### PR DESCRIPTION
**Pull request by smera**
This PR adds id/parent id information for each entry in the profiler report. This allows for future visualizers to display the report in a tree-like structure, where inclusive/exclusive times can be seen more clearly.

Additionally, it fixes a bug where entries representing the same element were not merged properly.

### CODE REVIEW

**Comment by smera** [comment]
This PR actually breaks the binary logger format. But since the last version update happened a couple of days ago, it might be fine. Please let me know, otherwise I'll bump the version and add the appropriate checks for back-compat.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 1]
```diff
@@ -841,6 +841,10 @@ private ProfiledLocation ReadProfiledLocation()
 
         private EvaluationLocation ReadEvaluationLocation()
         {
+            var id = ReadInt32();
+            int? parentId = null;
+            var hasParent = ReadBoolean();
+            if (hasParent) parentId = ReadInt32();
```
Comment (cdmihai) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs]: nit: add curlies for single line statement blocks

Comment (smera) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs]: I can change this, but I copied the style from below (look at hasLine). I thought that was the coding convention?

---
In reply to: [157254270](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157254270) [](ancestors = 157254270)

File: src/Framework/Profiler/EvaluationIdProvider.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        /// </summary>
+        /// <remarks>
+        /// The id is guaranteed to be unique across all running processes
+        /// </remarks>
+        public static int GetNextId()
+        {
+            var nextId = Interlocked.Increment(ref _sAssignedId);
+            // Returns a unique number based on nextId (a unique number for this process) and the current process Id
+            // Uses the Cantor pairing function (https://en.wikipedia.org/wiki/Pairing_function) to guarantee uniqueness
+            return (((nextId + ProcessId) * (nextId + ProcessId + 1)) / 2) + ProcessId;
```
Comment (cdmihai) [inline src/Framework/Profiler/EvaluationIdProvider.cs]: :thumbsup:

File: src/Framework/Profiler/EvaluationLocation.cs [hunk 3]
```diff
@@ -93,94 +99,135 @@ public struct EvaluationLocation
...
         /// <nodoc/>
-        public string Description { get; }
+        public string ElementDescription { get; }
 
         /// <nodoc/>
         public EvaluationLocationKind Kind { get; }
 
         /// <nodoc/>
-        public static EvaluationLocation CreateLocationForCondition(EvaluationPass evaluationPass, string evaluationDescription, string file,
+        public bool IsEvaluationPass => File == null;
```
Comment (cdmihai) [inline src/Framework/Profiler/EvaluationLocation.cs]: Hopefully this will always hold :)

**Comment by cdmihai** [comment]
@smera there's been a release in the meantime. Can you please bump up the version and branch in the reader? :)

File: ref/net46/Microsoft.Build.Framework/Microsoft.Build.Framework.cs [hunk 4]
```diff
@@ -547,22 +547,31 @@ public sealed partial class TelemetryEventArgs : Microsoft.Build.Framework.Build
 }
 namespace Microsoft.Build.Framework.Profiler
 {
+    public static partial class EvaluationIdProvider
```
Comment (cdmihai) [inline ref/net46/Microsoft.Build.Framework/Microsoft.Build.Framework.cs]: Can you make this private? I don't see a reason it being public.

File: ref/net46/Microsoft.Build.Framework/Microsoft.Build.Framework.cs [hunk 5]
```diff
@@ -571,12 +580,20 @@ public partial struct EvaluationLocation
         public Microsoft.Build.Framework.Profiler.EvaluationLocation WithFileLineAndCondition(string file, System.Nullable<int> line, string condition) { throw null; }
         public Microsoft.Build.Framework.Profiler.EvaluationLocation WithFileLineAndElement(string file, System.Nullable<int> line, Microsoft.Build.Framework.IProjectElement element) { throw null; }
         public Microsoft.Build.Framework.Profiler.EvaluationLocation WithGlob(string globDescription) { throw null; }
+        public Microsoft.Build.Framework.Profiler.EvaluationLocation WithParentId(System.Nullable<int> parentId) { throw null; }
+    }
+    public partial class EvaluationLocationIdAgnosticComparer : System.Collections.Generic.IEqualityComparer<Microsoft.Build.Framework.Profiler.EvaluationLocation>
```
Comment (cdmihai) [inline ref/net46/Microsoft.Build.Framework/Microsoft.Build.Framework.cs]: Make private.

File: src/Build/Logging/ProfilerLogger.cs [hunk 6]
```diff
@@ -96,59 +96,165 @@ private void ProjectEvaluationFinishedRaised(object sender, BuildEventArgs e)
...
+                aggregatedLocations[originalKey] = profiledLocation;
+                // And we update the merge map to flag that the merge happened
+                mergeMap[pairToMerge.Key.Id] = originalKey.Id;
+            }
+            else
+            {
+                // The item is new, so the profiled times are legit, nothing to aggregate
+                // But we have to check if this item points to a parent that got merged
+                int mergedParent;
+                if (mergeMap.TryGetValue(pairToMerge.Key.Id, out mergedParent))
```
Comment (cdmihai) [inline src/Build/Logging/ProfilerLogger.cs]: Shouldn't this be pairToMerge.Key.ParentId? You are remapping the parent, not the location itself

Comment (cdmihai) [reply src/Build/Logging/ProfilerLogger.cs]: Speaking of, are there tests hitting this?

---
In reply to: [157267787](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157267787) [](ancestors = 157267787)

Comment (cdmihai) [reply src/Build/Logging/ProfilerLogger.cs]: oops :)

---
In reply to: [157268080](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157268080) [](ancestors = 157268080,157267787)

Comment (cdmihai) [reply src/Build/Logging/ProfilerLogger.cs]: The merge map potentially forms a chain. Shouldn't you need to follow the chain and use the top parent? I think that only the top parent goes into the AggregatedLocations, right? N locations get merged to one.

---
In reply to: [157268268](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157268268) [](ancestors = 157268268,157268080,157267787)

Comment (smera) [reply src/Build/Logging/ProfilerLogger.cs]: They don't form a chain. An item can be merged multiple times, but they all get merged into the first occurrence of the item regarding Ids

---
In reply to: [157268972](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157268972) [](ancestors = 157268972,157268268,157268080,157267787)

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 1 above)
Comment (cdmihai) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs]: There's more curlied single lines in this file than non-curlied single lines. Curlied lines wins :)

---
In reply to: [157256344](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157256344) [](ancestors = 157256344,157254270)

File: src/Framework/Profiler/EvaluationLocation.cs (hunk 3 above)
Comment (smera) [reply src/Framework/Profiler/EvaluationLocation.cs]: That's the contract so far

---
In reply to: [157258488](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157258488) [](ancestors = 157258488)

File: ref/net46/Microsoft.Build.Framework/Microsoft.Build.Framework.cs (hunk 4 above)
Comment (smera) [reply ref/net46/Microsoft.Build.Framework/Microsoft.Build.Framework.cs]: Made internal

---
In reply to: [157260091](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157260091) [](ancestors = 157260091)

File: ref/net46/Microsoft.Build.Framework/Microsoft.Build.Framework.cs (hunk 5 above)
Comment (smera) [reply ref/net46/Microsoft.Build.Framework/Microsoft.Build.Framework.cs]: Moved to where the profiler logger is and made internal

---
In reply to: [157260451](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157260451) [](ancestors = 157260451)

File: src/Build/Logging/ProfilerLogger.cs (hunk 6 above)
Comment (smera) [reply src/Build/Logging/ProfilerLogger.cs]: The usual Target evaluation hits this one

---
In reply to: [157268004](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157268004) [](ancestors = 157268004,157267787)

**Comment by smera** [comment]
Updated, new commit on the way

---
In reply to: [352068502](https://github.com/Microsoft/msbuild/pull/2813#issuecomment-352068502) [](ancestors = 352068502)

**Comment by smera** [comment]
@micodoba@microsoft.com  Pushed, just in case you want to take a look

---
In reply to: [352148433](https://github.com/Microsoft/msbuild/pull/2813#issuecomment-352148433) [](ancestors = 352148433,352068502)

**Comment by smera** [comment]
@dotnet-bot test this please

**Comment by smera** [comment]
@dotnet-bot test this please

**Comment by rainersigwald** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please

**Comment by rainersigwald** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please

**Comment by smera** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please

**Pull request merged**